lib/Sema/TypeCheckGeneric.cpp

```diff
@@ -170,6 +171,113 @@ static void revertDependentTypeLoc(TypeLoc &tl) {
+    std::copy(outerGenericSignature->getRequirements().begin(),
+              outerGenericSignature->getRequirements().end(),
+              std::back_inserter(interfaceRequirements));
+  }
+
+  unsigned returnTypeDepth = 0;
+  if (!interfaceGenericParams.empty())
+    returnTypeDepth = interfaceGenericParams.back()->getDepth() + 1;
+  auto returnTypeParam = GenericTypeParamType::get(returnTypeDepth, 0,
+                                                   tc.Context);
+  interfaceGenericParams.push_back(returnTypeParam);
+
+  if (constraintType->getClassOrBoundGenericClass()) {
+    // Use the type as a superclass constraint.
+    interfaceRequirements.push_back(Requirement(RequirementKind::Superclass,
+                                              returnTypeParam, constraintType));
+  } else {
+    auto constraints = constraintType->getExistentialLayout();
+    if (auto superclass = constraints.getSuperclass()) {
+      interfaceRequirements.push_back(Requirement(RequirementKind::Superclass,
+                                                  returnTypeParam, superclass));
+    }
+    for (auto protocol : constraints.getProtocols()) {
+      interfaceRequirements.push_back(Requirement(RequirementKind::Conformance,
+                                                  returnTypeParam, protocol));
+    }
+    if (auto layout = constraints.getLayoutConstraint()) {
+      interfaceRequirements.push_back(Requirement(RequirementKind::Layout,
+                                                  returnTypeParam, layout));
+    }
+  }
+  
+  auto interfaceSignature = GenericSignature::get(interfaceGenericParams,
+                                                  interfaceRequirements);
+  
+  // Create the OpaqueTypeDecl for the result type.
+  // It has the same parent context and generic environment as the originating
+  // decl.
+  auto dc = originatingDecl->getDeclContext();
+  // TODO: generalize
+  auto genericParams = cast<FuncDecl>(originatingDecl)->getGenericParams();
+
+  auto opaqueDecl = new (tc.Context) OpaqueTypeDecl(originatingDecl,
+                                                    genericParams,
+                                                    dc,
+                                                    interfaceSignature,
+                                                    returnTypeParam);
+  opaqueDecl->copyFormalAccessFrom(originatingDecl);
+  if (auto originatingEnv = originatingDC->getGenericEnvironmentOfContext()) {
+    opaqueDecl->setGenericEnvironment(originatingEnv);
+  }
+
+  originatingDecl->setOpaqueResultTypeDecl(opaqueDecl);
+  
+  // The declared interface type is an opaque ArchetypeType.
+  SubstitutionMap subs;
+  if (outerGenericSignature) {
+    subs = outerGenericSignature->getIdentitySubstitutionMap();
+  }
+  auto opaqueTy = OpaqueTypeArchetypeType::get(opaqueDecl, subs);
+  auto metatype = MetatypeType::get(opaqueTy);
+  opaqueDecl->setInterfaceType(metatype);
```

Comment:
The code for setting the type here is the same as in deserialization. Take a look at how NominalTypeDecl and AssociatedTypeDecl have `computeType()` methods; OpaqueTypeDecl should also have one.